// kinvey-nativescript-sdk/src/popup/popup.android.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var frameModule = require("ui/frame");
var page_1 = require("ui/page");
var grid_layout_1 = require("ui/layouts/grid-layout");
var stack_layout_1 = require("ui/layouts/stack-layout");
var web_view_1 = require("ui/web-view");
var OAuthWebViewHelper = (function (_super) {
    __extends(OAuthWebViewHelper, _super);
    function OAuthWebViewHelper() {
        var _this = _super.call(this) || this;
        return global.__native(_this);
    }
    OAuthWebViewHelper.initWithWebViewAndIntercept = function (webView, webViewIntercept) {
        var webViewCreateNativeView = webView.createNativeView;
        webView.createNativeView = function () {
            webView._webViewClient = OAuthWebViewHelper.initWithView(webView, webViewIntercept);
            var nativeView = new android.webkit.WebView(webView._context);
            nativeView.getSettings().setJavaScriptEnabled(true);
            nativeView.getSettings().setBuiltInZoomControls(true);
            nativeView.setWebViewClient(webView._webViewClient);
            nativeView.client = webView._webViewClient;
            return nativeView;
        };
    };
    OAuthWebViewHelper.initWithView = function (view, webViewIntercept) {
        try {
            var client = new OAuthWebViewHelper();
            client._view = view;
            client._origClient = view._webViewClient;
            client._webViewIntercept = webViewIntercept;
            return client;
        }
        catch (error) {
            console.dir(error);
        }
    };
    OAuthWebViewHelper.prototype.shouldOverrideUrlLoading = function (view, url) {
        var urlStr = '';
        if (typeof url === 'string') {
            urlStr = url;
        }
        else if (typeof url === 'object') {
            try {
                urlStr = url.getUrl().toString();
            }
            catch (ex) {
                return false;
            }
        }
        else {
            return false;
        }
        if (this._webViewIntercept(this._view, null, urlStr)) {
            return true;
        }
        return false;
    };
    OAuthWebViewHelper.prototype.onPageStarted = function (view, url, favicon) {
        this._webViewIntercept(this._view, null, url);
        _super.prototype.onPageStarted.call(this, view, url, favicon);
        if (this._view) {
            this._view._onLoadStarted(url, undefined);
        }
    };
    OAuthWebViewHelper.prototype.onPageFinished = function (view, url) {
        _super.prototype.onPageFinished.call(this, view, url);
        if (this._view) {
            this._webViewIntercept(this._view, null, url);
            this._view._onLoadFinished(url, undefined);
        }
    };
    OAuthWebViewHelper.prototype.onReceivedError = function () {
        var view = arguments[0];
        if (arguments.length === 4) {
            var errorCode = arguments[1];
            var description = arguments[2];
            var failingUrl = arguments[3];
            this._webViewIntercept(this._view, null, failingUrl);
            _super.prototype.onReceivedError.call(this, view, errorCode, description, failingUrl);
            if (this._view) {
                this._view._onLoadFinished(failingUrl, description + "(" + errorCode + ")");
            }
        }
        else {
            var request = arguments[1];
            var error = arguments[2];
            this._webViewIntercept(this._view, error);
            _super.prototype.onReceivedError.call(this, view, request, error);
            if (this._view) {
                this._view._onLoadFinished(error.getUrl && error.getUrl(), error.getDescription() + "(" + error.getErrorCode() + ")");
            }
        }
    };
    return OAuthWebViewHelper;
}(android.webkit.WebViewClient));
var OAuthPageProvider = (function () {
    function OAuthPageProvider(authUrl, webViewIntercept) {
        this._authUrl = authUrl;
        this._webViewIntercept = webViewIntercept;
    }
    OAuthPageProvider.prototype.createWebViewPage = function () {
        var webView = new web_view_1.WebView();
        OAuthWebViewHelper.initWithWebViewAndIntercept(webView, this._webViewIntercept);
        var grid = new grid_layout_1.GridLayout();
        grid.addChild(webView);
        var stack = new stack_layout_1.StackLayout();
        stack.addChild(grid);
        var page = new page_1.Page();
        page.content = stack;
        webView.src = this._authUrl;
        return page;
    };
    return OAuthPageProvider;
}());
var Popup = (function (_super) {
    __extends(Popup, _super);
    function Popup() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._open = false;
        return _this;
    }
    Popup.prototype.open = function (url) {
        var _this = this;
        if (url === void 0) { url = '/'; }
        if (this._open === false) {
            var webViewIntercept = function (webView, error, url) {
                var urlStr = '';
                try {
                    if (error && error.userInfo && error.userInfo.allValues && error.userInfo.allValues.count > 0) {
                        var val0 = error.userInfo.allValues[0];
                        if (val0.absoluteString) {
                            urlStr = val0.absoluteString;
                        }
                        else if (val0.userInfo && val0.userInfo.allValues && val0.userInfo.allValues.count > 0) {
                            urlStr = val0.userInfo.allValues[0];
                        }
                        else {
                            urlStr = val0;
                        }
                    }
                    else if (webView.request && webView.request.URL && webView.request.URL.absoluteString) {
                        urlStr = webView.request.URL.absoluteString;
                    }
                    else if (url) {
                        urlStr = url;
                    }
                }
                catch (ex) {
                }
                _this.emit('loadstop', { url: urlStr });
                return true;
            };
            var authPage_1 = new OAuthPageProvider(url, webViewIntercept);
            frameModule.topmost().navigate(function () { return authPage_1.createWebViewPage(); });
            this._open = true;
        }
        return this;
    };
    Popup.prototype.close = function () {
        if (this._open) {
            frameModule.topmost().goBack();
            this._open = false;
        }
        return this;
    };
    return Popup;
}(events_1.EventEmitter));
exports.Popup = Popup;
